src/road_gui.cpp

```diff
@@ -304,13 +304,16 @@ struct BuildRoadToolbarWindow : Window {
 	{
 		if (!gui_scope) return;
 
-		if (_game_mode != GM_EDITOR && !CanBuildVehicleInfrastructure(VEH_ROAD, GetRoadTramType(this->roadtype))) delete this;
 		bool can_build = _game_mode != GM_EDITOR;
-		this->SetWidgetsDisabledState(!can_build,
-				WID_ROT_DEPOT,
-				WID_ROT_BUS_STATION,
-				WID_ROT_TRUCK_STATION,
-				WIDGET_LIST_END);
+		if (can_build && !CanBuildVehicleInfrastructure(VEH_ROAD, GetRoadTramType(this->roadtype))) {
+			delete this;
+		} else {
+			this->SetWidgetsDisabledState(!can_build,
+					WID_ROT_DEPOT,
+					WID_ROT_BUS_STATION,
+					WID_ROT_TRUCK_STATION,
+					WIDGET_LIST_END);
+		}
```

Comment:
The `else` can be removed
```suggestion
		if (can_build && !CanBuildVehicleInfrastructure(VEH_ROAD, GetRoadTramType(this->roadtype))) {
			delete this;
			return;
		}
		this->SetWidgetsDisabledState(!can_build,
				WID_ROT_DEPOT,
				WID_ROT_BUS_STATION,
				WID_ROT_TRUCK_STATION,
				WIDGET_LIST_END);
```

Reply:
I think this `SetWidgetsDisabledState` call could be removed as well. There are 2 toolbars, one for the scenario editor and one for the normal game. The scenario edition does not even contain the depot and station widgets. And the `can_build` depends on only whether the game mode is scenario editor or not. It is possibly a remnant from an earlier version when the scenario editor did not have its own toolbar.

https://github.com/OpenTTD/OpenTTD/blob/71e79edfc1adaa4ea1e19d6ec87cb045ef7f69fe/src/road_gui.cpp#L788

https://github.com/OpenTTD/OpenTTD/blob/71e79edfc1adaa4ea1e19d6ec87cb045ef7f69fe/src/road_gui.cpp#L890

And because of this, maybe the delete windows block could be also removed.

Reply:
No the `SetWidgetsDisabledState` call is needed for the hotkeys, even if the widgets are not shown in scenario editor, they are still existing and could be triggered via hotkeys if not disabled.

Reply:
I removed those lines.